### DISCUSSION TITLE
Python ScanTailor [WIP] - Advice and Contributions Welcome

### DISCUSSION THREAD

**Discussion by adehad** [ScanTailor-Advanced/scantailor-advanced#87] [Show and tell]
Hello!

ScanTailor has provided much for the community, and in my own way of contributing back I have started a Python port: https://github.com/adehad/py-scantailor-advanced .

Definitely may be asked why I didn't choose X framework or go in Y direction.

But in brief I'm hoping to reduce the barrier to entry for contributions.
1. **New contributors**: are more likely to have a bit of Python experience to apply to specific problems
2. **Previous contributors**: can leverage the familiarity with the PySide6 (Python bindings for Qt), but hopefully with added flexibility from the higher level abstractions and dependency management (e.g. can install a dependency and use more trivially - say opencv)

I'm hoping that my making this post:
1. **New contributors**: might be able to start contributing to the new port
2. **Previous contributors**: can provide any advice from this project and what should be considered/changed in such a porting exercise, contributing too if they like

The current state of the affairs of the port, as of writing, is that:
1. project structure is approximately in place
2. devcontainer is setup (I am running on Windows and can get the MainWindow opening, as well as other Qt tools eg QtDesigner)
3. debugging is available
4. test framework is available, sample test in place
5. no project functionality is in place, currently the viewing of all ui files might be possible, but that is all.

With 60k lines of code and 30k lines of translations it is no small task, but hopefully will be a fruitful project in the long term

**Comment by FriedrichFroebel** [comment]
While it is great to see someone tackling the future of the project, the tasks you are planning to do here will probably take quite some time until feature-parity is reached.

Some aspects I stumbled upon when having a quick look at the project:

* Forks have Issues and Discussions disabled by default. Enabling them might help you find further supporters.
* Forks are not watched by you by default, thus notifications tend to get lost. You might want to enable them.
* Considering common Ubuntu 22.04 LTS installations, limiting this to Python 3.12 directly (as well as Qt 6 only) might impose too much restrictions.
* There probably should be a clear plan of which steps are planned in which order.

**Reply by adehad** [reply]
Thanks for the targetted feedback. Really appreciate it given, as we are aware, the scale of the project.

Regarding the Ubuntu/Python/Qt combination, this is managed through the Dockerfile setup, with the devcontainer configuration the aim is it is sufficient for a developer to have: Docker and VSCode, and that will be enough. 
- Technically they can even start a codespace in the web browser, but won't be able to preview the UI, but could easily work on algorithm implementations etc.
- Naturally users of other preferred developer workflows can create their own Dockerfile with the project's as a base, installing any relevant tools. 


Completely agree on the plan, I've enabled the Project feature, and will need to spend some time establishing the priorities of the project. Welcome any suggestions/advice
